# Lesson 2

## Working with requests library

Now we will start doing a request to the following websites: Google, NBA and Rotten Tomatoes. 

Let's check their status_code by using the requests method GET, which simply indicates that you're trying to retrieve information from the data source:

In [1]:
import requests

google = requests.get("https://developers.google.com")
print("Google:", google.status_code)

NBA = response = requests.get("https://api.sportsdata.io/api/nba/fantasy/json/CurrentSeason")
print("NBA:", NBA.status_code)

rotten_tomato = requests.get("http://api.rottentomatoes.com/api/public/v1.0/lists/movies/box_office.json")
print("Rotten Tomatoes:", rotten_tomato.status_code)

Google: 200
NBA: 401
Rotten Tomatoes: 403


## Status codes

So we can see we receive some status codes, the most frequent are these:

* **200**: Everything went okay and the result has been returned (if any).

* **301**: The server is redirecting you to a different endpoint. This can happen when a company switches domain names, or an endpoint name is changed.

* **400**: The server thinks you made a bad request. This happens when you don’t send along the right data, among other things.

* **401**: You are not properly authenticated.

* **403**: The resource you’re trying to access is forbidden: you don’t have the right permissions to get it.

* **404**: The resource you tried to access doesn't exist.

* **503**: The server can't handle the request.

## After connecting

As we are connected now to the google developer endpoint let's try to get some data from it.

In [2]:
# print(google.text)

This is not the proper format for us right now so let's try to tranform it to the one we need, the **JSON (JavaScript Object Notation) format**, which is the language of RESTful APIs. It is the way to encode data to make data easily readable by machines.

In [3]:
# google.json()

The problem here is that we are not connected to a proper API that has a json as the endpoint, so we will need to get another valid URL. Remember that the code 200 means that the connection has been established, but not that it is the connection you need. Let's try with the skyscrapper API .

We will connect to this API through the web [RapidAPI ](https://rapidapi.com/skyscanner/api/skyscanner-flight-search)so you will need to create an account to get a personal Key (user ID) to have access to it, just sign up and get your Key. The "problem" with this API is that we will need to pass some parameters, including the headers and the city following this code:

In [4]:
import sys
import import_ipynb
# sys.path.insert(1, '/content/drive/MyDrive/Credentials/')
from config import *
import requests
# import config

importing Jupyter notebook from config.ipynb


In [5]:
# airport code lookup
# takes first one in the json response for airport location lookupand returns it

def lookup_city_code(city_name):
    
    url = "https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/autosuggest/v1.0/UK/GBP/en-GB/"
    params = {"query":city_name}
    
    headers = {'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
                      'x-rapidapi-key': x_rapidapi_key}

    response = requests.get(url, headers = headers, params = params)
    
    return response.json()['Places'][0]['PlaceId']


In [6]:
lookup_city_code('London')

'LOND-sky'

In [7]:
url = "https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/autosuggest/v1.0/UK/GBP/en-GB/"

params = {"query":"NRT"}

#headers = {'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
#                      'x-rapidapi-key': "<introduce your RapidAPI key here>"}

headers = {'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
                      'x-rapidapi-key': x_rapidapi_key}

# Notice the additional keywords here (headers, params)
response = requests.get(url, headers = headers, params = params)

response.json()

{'Places': [{'PlaceId': 'NRT-sky',
   'PlaceName': 'Tokyo Narita',
   'CountryId': 'JP-sky',
   'RegionId': '',
   'CityId': 'TYOA-sky',
   'CountryName': 'Japan'},
  {'PlaceId': 'NOTT-sky',
   'PlaceName': 'Nottingham',
   'CountryId': 'UK-sky',
   'RegionId': '',
   'CityId': 'NOTT-sky',
   'CountryName': 'United Kingdom'},
  {'PlaceId': 'EMA-sky',
   'PlaceName': 'East Midlands',
   'CountryId': 'UK-sky',
   'RegionId': '',
   'CityId': 'NOTT-sky',
   'CountryName': 'United Kingdom'},
  {'PlaceId': 'NQT-sky',
   'PlaceName': 'Nottingham',
   'CountryId': 'UK-sky',
   'RegionId': '',
   'CityId': 'NOTT-sky',
   'CountryName': 'United Kingdom'},
  {'PlaceId': 'NL-sky',
   'PlaceName': 'Netherlands',
   'CountryId': 'NL-sky',
   'RegionId': '',
   'CityId': '-sky',
   'CountryName': 'Netherlands'},
  {'PlaceId': 'CCU-sky',
   'PlaceName': 'Netaji Subhash Chandra Bose International',
   'CountryId': 'IN-sky',
   'RegionId': '',
   'CityId': 'ICCU-sky',
   'CountryName': 'India'},
  {'Pl

You should get:

{"Places":[{"PlaceId":"TYOA-sky","PlaceName":"Tokyo","CountryId":"JP-sky","RegionId":"","CityId":"TYOA-sky","CountryName":"Japan"},{"PlaceId":"NRT-sky","PlaceName":"Tokyo Narita","CountryId":"JP-sky","RegionId":"","CityId":"TYOA-sky","CountryName":"Japan"},{"PlaceId":"HND-sky","PlaceName":"Tokyo Haneda","CountryId":"JP-sky","RegionId":"","CityId":"TYOA-sky","CountryName":"Japan"},{"PlaceId":"TJH-sky","PlaceName":"Toyooka","CountryId":"JP-sky","RegionId":"","CityId":"JTJH-sky","CountryName":"Japan"},{"PlaceId":"OOK-sky","PlaceName":"Toksook Bay","CountryId":"US-sky","RegionId":"AK","CityId":"OOKA-sky","CountryName":"United States"},{"PlaceId":"TKZ-sky","PlaceName":"Tokoroa","CountryId":"NZ-sky","RegionId":"","CityId":"TKZN-sky","CountryName":"New Zealand"}]}

To know the parameters we need to be able to connect properly to an API we should look first at its documentation. For the next query we will look at the prices for the flights from San Francisco to New York City in 2020/12/12

In [8]:
url = "https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/browsequotes/v1.0/US/USD/en-US/SFO-sky/NYCA-sky/2020-12-12"

params = {"inboundpartialdate":"2020-12-12"}

#headers = {
#    'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
#    'x-rapidapi-key': "<introduce your RapidAPI key here>"}

headers = {
    'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
    'x-rapidapi-key': x_rapidapi_key}    

response = requests.get(url, headers=headers, params=params)

response.json()

{'Quotes': [{'QuoteId': 1,
   'MinPrice': 43,
   'Direct': False,
   'OutboundLeg': {'CarrierIds': [1065],
    'OriginId': 81727,
    'DestinationId': 50290,
    'DepartureDate': '2020-12-12T00:00:00'},
   'QuoteDateTime': '2020-11-28T05:48:00'},
  {'QuoteId': 2,
   'MinPrice': 79,
   'Direct': True,
   'OutboundLeg': {'CarrierIds': [851],
    'OriginId': 81727,
    'DestinationId': 50290,
    'DepartureDate': '2020-12-12T00:00:00'},
   'QuoteDateTime': '2020-11-28T05:48:00'}],
 'Carriers': [{'CarrierId': 1065, 'Name': 'Frontier Airlines'},
  {'CarrierId': 851, 'Name': 'Alaska Airlines'}],
 'Places': [{'Name': 'New York Newark',
   'Type': 'Station',
   'PlaceId': 50290,
   'IataCode': 'EWR',
   'SkyscannerCode': 'EWR',
   'CityName': 'New York',
   'CityId': 'NYCA',
   'CountryName': 'United States'},
  {'Name': 'San Francisco International',
   'Type': 'Station',
   'PlaceId': 81727,
   'IataCode': 'SFO',
   'SkyscannerCode': 'SFO',
   'CityName': 'San Francisco',
   'CityId': 'SFOA'

The output will be:

{'Quotes': [{'QuoteId': 1,
   'MinPrice': 92.0,
   'Direct': False,
   'OutboundLeg': {'CarrierIds': [1065],
    'OriginId': 81727,
    'DestinationId': 50290,
    'DepartureDate': '2020-12-12T00:00:00'},
   'QuoteDateTime': '2020-09-21T10:49:00'},
  {'QuoteId': 2,
   'MinPrice': 133.0,
   'Direct': True,
   'OutboundLeg': {'CarrierIds': [851],
    'OriginId': 81727,
    'DestinationId': 50290,
    'DepartureDate': '2020-12-12T00:00:00'},
   'QuoteDateTime': '2020-09-21T10:49:00'}],
 'Places': [{'PlaceId': 50290,
   'IataCode': 'EWR',
   'Name': 'New York Newark',
   'Type': 'Station',
   'SkyscannerCode': 'EWR',
   'CityName': 'New York',
   'CityId': 'NYCA',
   'CountryName': 'United States'},
  {'PlaceId': 60987,
   'IataCode': 'JFK',
   'Name': 'New York John F. Kennedy',
   'Type': 'Station',
   'SkyscannerCode': 'JFK',
   'CityName': 'New York',
   'CityId': 'NYCA',
   'CountryName': 'United States'},
  {'PlaceId': 65633,
   'IataCode': 'LGA',
   'Name': 'New York LaGuardia',
   'Type': 'Station',
   'SkyscannerCode': 'LGA',
   'CityName': 'New York',
   'CityId': 'NYCA',
   'CountryName': 'United States'},
  {'PlaceId': 81727,
   'IataCode': 'SFO',
   'Name': 'San Francisco International',
   'Type': 'Station',
   'SkyscannerCode': 'SFO',
   'CityName': 'San Francisco',
   'CityId': 'SFOA',
   'CountryName': 'United States'}],
 'Carriers': [{'CarrierId': 819, 'Name': 'Aegean Airlines'},
  {'CarrierId': 851, 'Name': 'Alaska Airlines'},
  {'CarrierId': 870, 'Name': 'jetBlue'},
  {'CarrierId': 1065, 'Name': 'Frontier Airlines'},
  {'CarrierId': 1721, 'Name': 'Sun Country Airlines'},
  {'CarrierId': 1793, 'Name': 'United'},
  {'CarrierId': 1902, 'Name': 'Southwest Airlines'}],
 'Currencies': [{'Code': 'USD',
   'Symbol': '$',
   'ThousandsSeparator': ',',
   'DecimalSeparator': '.',
   'SymbolOnLeft': True,
   'SpaceBetweenAmountAndSymbol': False,
   'RoundingCoefficient': 0,
   'DecimalDigits': 2}]}

# Lesson 3

## Handling JSON files

In [9]:
# Import libraries
import pandas as pd
import json

In [10]:
my_response = {'Quotes': [{'QuoteId': 1,
   'MinPrice': 92.0,
   'Direct': False,
   'OutboundLeg': {'CarrierIds': [1065],
    'OriginId': 81727,
    'DestinationId': 50290,
    'DepartureDate': '2020-12-12T00:00:00'},
   'QuoteDateTime': '2020-09-21T10:49:00'},
  {'QuoteId': 2,
   'MinPrice': 133.0,
   'Direct': True,
   'OutboundLeg': {'CarrierIds': [851],
    'OriginId': 81727,
    'DestinationId': 50290,
    'DepartureDate': '2020-12-12T00:00:00'},
   'QuoteDateTime': '2020-09-21T10:49:00'}],
 'Places': [{'PlaceId': 50290,
   'IataCode': 'EWR',
   'Name': 'New York Newark',
   'Type': 'Station',
   'SkyscannerCode': 'EWR',
   'CityName': 'New York',
   'CityId': 'NYCA',
   'CountryName': 'United States'},
  {'PlaceId': 60987,
   'IataCode': 'JFK',
   'Name': 'New York John F. Kennedy',
   'Type': 'Station',
   'SkyscannerCode': 'JFK',
   'CityName': 'New York',
   'CityId': 'NYCA',
   'CountryName': 'United States'},
  {'PlaceId': 65633,
   'IataCode': 'LGA',
   'Name': 'New York LaGuardia',
   'Type': 'Station',
   'SkyscannerCode': 'LGA',
   'CityName': 'New York',
   'CityId': 'NYCA',
   'CountryName': 'United States'},
  {'PlaceId': 81727,
   'IataCode': 'SFO',
   'Name': 'San Francisco International',
   'Type': 'Station',
   'SkyscannerCode': 'SFO',
   'CityName': 'San Francisco',
   'CityId': 'SFOA',
   'CountryName': 'United States'}],
 'Carriers': [{'CarrierId': 819, 'Name': 'Aegean Airlines'},
  {'CarrierId': 851, 'Name': 'Alaska Airlines'},
  {'CarrierId': 870, 'Name': 'jetBlue'},
  {'CarrierId': 1065, 'Name': 'Frontier Airlines'},
  {'CarrierId': 1721, 'Name': 'Sun Country Airlines'},
  {'CarrierId': 1793, 'Name': 'United'},
  {'CarrierId': 1902, 'Name': 'Southwest Airlines'}],
 'Currencies': [{'Code': 'USD',
   'Symbol': '$',
   'ThousandsSeparator': ',',
   'DecimalSeparator': '.',
   'SymbolOnLeft': True,
   'SpaceBetweenAmountAndSymbol': False,
   'RoundingCoefficient': 0,
   'DecimalDigits': 2}]}

In [11]:
pd.json_normalize(my_response)
# If this doesn't work, try to import json_normalize directly from pandas.io.json

Quotes  \
0  [{'QuoteId': 1, 'MinPrice': 92.0, 'Direct': Fa...   

                                              Places  \
0  [{'PlaceId': 50290, 'IataCode': 'EWR', 'Name':...   

                                            Carriers  \
0  [{'CarrierId': 819, 'Name': 'Aegean Airlines'}...   

                                          Currencies  
0  [{'Code': 'USD', 'Symbol': '$', 'ThousandsSepa...

As an output we have a DataFrame with 4 columns, Quotes, Places, Carriers and Currencies, but the values will also be in JSON format. From this we can infere that we are not working with a dataset, but a relational database, and the columns are the datasets we were looking for. Let's transform them into pd.DataFrames:

In [12]:
quotes = pd.DataFrame(pd.json_normalize(my_response)["Quotes"][0])
carriers = pd.DataFrame(pd.json_normalize(my_response)["Carriers"][0])
places = pd.DataFrame(pd.json_normalize(my_response)["Places"][0])
currencies = pd.DataFrame(pd.json_normalize(my_response)["Currencies"][0])

print("Quotes DataFrame extract: ")
display(quotes.head())
print()
print("Carriers DataFrame extract: ")
display(carriers.head())
print()
print("Places DataFrame extract: ")
display(places.head())
print()
print("Currencies DataFrame extract: ")
display(currencies.head())

Quotes DataFrame extract: 


QuoteId  MinPrice  Direct  \
0        1      92.0   False   
1        2     133.0    True   

                                         OutboundLeg        QuoteDateTime  
0  {'CarrierIds': [1065], 'OriginId': 81727, 'Des...  2020-09-21T10:49:00  
1  {'CarrierIds': [851], 'OriginId': 81727, 'Dest...  2020-09-21T10:49:00


Carriers DataFrame extract: 


CarrierId                  Name
0        819       Aegean Airlines
1        851       Alaska Airlines
2        870               jetBlue
3       1065     Frontier Airlines
4       1721  Sun Country Airlines


Places DataFrame extract: 


PlaceId IataCode                         Name     Type SkyscannerCode  \
0    50290      EWR              New York Newark  Station            EWR   
1    60987      JFK     New York John F. Kennedy  Station            JFK   
2    65633      LGA           New York LaGuardia  Station            LGA   
3    81727      SFO  San Francisco International  Station            SFO   

        CityName CityId    CountryName  
0       New York   NYCA  United States  
1       New York   NYCA  United States  
2       New York   NYCA  United States  
3  San Francisco   SFOA  United States


Currencies DataFrame extract: 


Code Symbol ThousandsSeparator DecimalSeparator  SymbolOnLeft  \
0  USD      $                  ,                .          True   

   SpaceBetweenAmountAndSymbol  RoundingCoefficient  DecimalDigits  
0                        False                    0              2

After reviewing and printing our data we can conclude that quotes are the flights we have for our selected origin and destination (just one in this case), places are the airports we could travel from and to, carriers are the companies usually have the kind of trip we have chosen, and currency is self-explanatory.

However we still have another json in one of the datasets, in Quotes, variable ["OutboundLeg"], that's due to the same principle we have been working for during this lesson, relational databases. Let's unfold that variable:

In [13]:
flights = pd.DataFrame(pd.DataFrame(pd.json_normalize(my_response)["Quotes"][0])["OutboundLeg"][0])
flights.head()

CarrierIds  OriginId  DestinationId        DepartureDate
0        1065     81727          50290  2020-12-12T00:00:00

# Lesson 4

Now that we know what JSON is and we have connected to the skyscanner API let's work with it. In this lesson we will create a function for which if we introduce two dates, an origin and a destination we will get the data from the cheapest flight between these two dates.

We will start by creating a function that takes two dates and return a list of dates between the given ones. Luckily, there is a pandas method for that.

In [22]:
def get_dates(start, end):
    return pd.Series(pd.date_range(start, end,freq='d').format())

get_dates('1992-01-01','1992-01-20')

0     1992-01-01
1     1992-01-02
2     1992-01-03
3     1992-01-04
4     1992-01-05
5     1992-01-06
6     1992-01-07
7     1992-01-08
8     1992-01-09
9     1992-01-10
10    1992-01-11
11    1992-01-12
12    1992-01-13
13    1992-01-14
14    1992-01-15
15    1992-01-16
16    1992-01-17
17    1992-01-18
18    1992-01-19
19    1992-01-20
dtype: object

In [15]:
def city_code(city_name):
    url = "https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/autosuggest/v1.0/US/USD/en-US/"

    params = {"query": city_name}

    headers = {'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
                      'x-rapidapi-key': x_rapidapi_key}

    response = requests.get(url, headers = headers, params = params)

    return response.json()["Places"][0]["PlaceId"]

In [16]:
city_code('London')

'LOND-sky'

In [17]:
# exercise
def flight_prices(departure, arrival, date):
#     date in format 2020-12-10
    departure_code = city_code(departure)
    arrival_code = city_code(arrival)
    
    url = f"https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/browsequotes/v1.0/US/USD/en-US/{departure_code}/{arrival_code}/{date}"    
    
    headers = {
    'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
    'x-rapidapi-key': x_rapidapi_key}

    response = requests.get(url, headers=headers, params=params)
    
    return response.json()
    

In [18]:
flight_prices('Paris', 'Shanghai','2020-12-10')

{'Quotes': [{'QuoteId': 1,
   'MinPrice': 460,
   'Direct': False,
   'OutboundLeg': {'CarrierIds': [1717],
    'OriginId': 44759,
    'DestinationId': 75575,
    'DepartureDate': '2020-12-10T00:00:00'},
   'QuoteDateTime': '2020-11-25T19:31:00'}],
 'Carriers': [{'CarrierId': 1717, 'Name': 'Aeroflot'}],
 'Places': [{'Name': 'Paris Charles de Gaulle',
   'Type': 'Station',
   'PlaceId': 44759,
   'IataCode': 'CDG',
   'SkyscannerCode': 'CDG',
   'CityName': 'Paris',
   'CityId': 'PARI',
   'CountryName': 'France'},
  {'Name': 'Shanghai Pu Dong',
   'Type': 'Station',
   'PlaceId': 75575,
   'IataCode': 'PVG',
   'SkyscannerCode': 'PVG',
   'CityName': 'Shanghai',
   'CityId': 'CSHA',
   'CountryName': 'China'},
  {'Name': 'Paris',
   'Type': 'City',
   'PlaceId': 3579769,
   'SkyscannerCode': 'PARI',
   'CityId': 'PARI',
   'CountryName': 'France'}],
 'Currencies': [{'Code': 'USD',
   'Symbol': '$',
   'ThousandsSeparator': ',',
   'DecimalSeparator': '.',
   'SymbolOnLeft': True,
   'S

Let's use the funtion to get the flights between two destinations on a given date.

In [19]:
# def flight_prices(departure, arrival, date):

#     departure_code = city_code(departure)
#     arrival_code = city_code(arrival)
#     url = f"https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/browsequotes/v1.0/US/USD/en-US/{departure_code}/{arrival_code}/{date}"

#     params = {"inboundpartialdate":{date}}

#     #headers = {
#     #'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
#     #'x-rapidapi-key': "<introduce your RapidAPI key here>"}

#     headers = {
#     'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
#     'x-rapidapi-key': x_rapidapi_key}

#     response = requests.get(url, headers=headers, params=params)

#     return response.json()

# flight_prices('Paris', 'London','2020-12-10')

Now that we already have our date function let's create another function that takes the origin, destination and dates as arguments. It will create a dictionary that has every day between the two dates and containts a call to the API with all the information about the days:

In [21]:
from  time import sleep

def flight_days(origin, destination, start, end):
    
    dates = get_dates(start, end)
#     gets dates above

    my_dict = {}
    
    for date in list(dates):
        
#     result = None
#     while (result == None):

        try: 
            result = flight_prices(origin, destination, date)
#           result is not none break loop, if none, keep trying  
            my_dict[date] = result

        except: 
            print("Error from the API")
            sleep(60)
                
    return my_dict

# everyday will be a key, value is number of flights for this day

# flight_days('Paris','London','2020-11-28','2020-12-10')

Error from the API
Error from the API


{'2020-11-27': {'errors': ['path param outboundDate Invalid Date Range provided 2020-11-27']},
 '2020-11-28': {'Quotes': [{'QuoteId': 1,
    'MinPrice': 131,
    'Direct': False,
    'OutboundLeg': {'CarrierIds': [1218],
     'OriginId': 73097,
     'DestinationId': 65698,
     'DepartureDate': '2020-11-28T00:00:00'},
    'QuoteDateTime': '2020-11-26T18:31:00'},
   {'QuoteId': 2,
    'MinPrice': 225,
    'Direct': True,
    'OutboundLeg': {'CarrierIds': [838],
     'OriginId': 44759,
     'DestinationId': 65698,
     'DepartureDate': '2020-11-28T00:00:00'},
    'QuoteDateTime': '2020-11-27T01:31:00'}],
  'Carriers': [{'CarrierId': 838, 'Name': 'Air France'},
   {'CarrierId': 1218, 'Name': 'Iberia'}],
  'Places': [{'Name': 'Paris Charles de Gaulle',
    'Type': 'Station',
    'PlaceId': 44759,
    'IataCode': 'CDG',
    'SkyscannerCode': 'CDG',
    'CityName': 'Paris',
    'CityId': 'PARI',
    'CountryName': 'France'},
   {'Name': 'London Heathrow',
    'Type': 'Station',
    'PlaceId'

Now let's create another function that returns the date of the cheapest flight.

In [141]:
def lowest_price(origin, destination, start, end):
    
    flights = flight_days(origin, destination, start, end)

    my_dict = {}
    for date in flights:   
        
        my_dict[date] = []
        
        for elem in flights[date]['Quotes']:

            my_dict[date].append(elem['MinPrice'])
            
            print("The min price for this date is: ",
                  min(my_dict[date]))

        #  my_dict[date] = min(my_dict[date]) 

    print(my_dict.values())
    return min(min(my_dict.values()))

The problem with this code is that it just gets the first date as the only one even if there are other days with the same price. Try to fix that in the lab.

In [ ]:
lowest_price('Barcelona','London','2020-11-26','2020-12-10')

In [ ]:
# lab 8.04
# create a function that returns the price, names of origin and arrival airports 
# and the name of the company. Do it for all the flights between two dates that cost the same.

In [76]:
def get_lowest_dates(origin, destination, start, end):
    
    flights = flight_days(origin, destination, start, end)

    my_dict = {}
    for date in flights:   
        
        my_dict[date] = []
        
        for elem in flights[date]['Quotes']:

            my_dict[date].append(elem['MinPrice'])
        
    lowest_price = min(min(my_dict.values()))    
    lowest_dates = []
        
    print(my_dict)    
    for key, value in my_dict.items():
        
        if any(v == lowest_price for v in value):
            lowest_dates.append(key)
    
    return lowest_dates  

get_lowest_dates('Barcelona','London','2020-11-28','2020-12-10')

{'2020-11-28': [139], '2020-11-29': [101, 258], '2020-11-30': [105, 175], '2020-12-01': [110], '2020-12-02': [91, 190], '2020-12-03': [107, 221], '2020-12-04': [94, 234], '2020-12-05': [124, 228], '2020-12-06': [66, 148], '2020-12-07': [74, 171], '2020-12-08': [70, 124], '2020-12-09': [70, 89], '2020-12-10': [77, 156]}


['2020-12-06']